File: ruby-watir-tests/spec/pages/welcome_page.rb
# Page Object Model - Welcome Page (after successful login)
# Encapsulates all interactions with the welcome/dashboard page
class WelcomePage
  attr_reader :browser

  def initialize(browser)
    @browser = browser
  end

  # ---- Element Accessors ----
  def welcome_title
    browser.h1(id: "welcome-title")
  end

  def welcome_message
    browser.element(id: "welcome-message")
  end

  def logout_button
    browser.button(id: "logout-button")
  end

  # ---- Actions ----
  def click_logout
    logout_button.click
    self
  end

  # ---- Assertions / Queries ----
  def displayed?
    browser.wait_until(timeout: 5) { welcome_title.present? }
    welcome_title.present? && logout_button.present?
  end

  def title_text
    welcome_title.text
  end

  def message_text
    welcome_message.text
  end

  def logged_in_as?(username)
    message_text.include?(username)
  end
end

File: playwright-tests/tests/pages/welcome-page.ts
import { Page, Locator, expect } from "@playwright/test";

/**
 * Page Object Model - Welcome Page (after successful login)
 * Encapsulates all interactions with the welcome/dashboard page
 */
export class WelcomePage {
  readonly page: Page;
  readonly welcomeTitle: Locator;
  readonly welcomeMessage: Locator;
  readonly logoutButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.welcomeTitle = page.locator("#welcome-title");
    this.welcomeMessage = page.locator("#welcome-message");
    this.logoutButton = page.locator("#logout-button");
  }

  // ---- Actions ----
  async clickLogout() {
    await this.logoutButton.click();
  }

  // ---- Assertions / Queries ----
  async isDisplayed() {
    await expect(this.welcomeTitle).toBeVisible({ timeout: 5000 });
    await expect(this.logoutButton).toBeVisible();
  }

  async getTitleText(): Promise<string> {
    return await this.welcomeTitle.innerText();
  }

  async getMessageText(): Promise<string> {
    return await this.welcomeMessage.innerText();
  }

  async isLoggedInAs(username: string): Promise<boolean> {
    const text = await this.getMessageText();
    return text.includes(username);
  }
}
